File: 67.cpp
#include <iostream>
#include <string>
#include <algorithm> //reverse()
#include <cassert>

using namespace std;

class Solution
{
public:
    string addBinary(string a, string b) 
    {
        //The input strings are both non-empty
        int lena = a.length();
        int lenb = b.length();
        string s;
        
        int carry = 0;
        int vala = 0, valb = 0;
		int tmp = 0;
        while (lena > 0 || lenb > 0)
        {
            --lena;
            --lenb;
            if (lena >= 0)
                vala = a[lena] - '0';
            else
                vala = 0;
            if (lenb >= 0)
                valb = b[lenb] - '0';
            else
                valb = 0;
            tmp = vala + valb + carry;
			if (tmp == 0)
			{
				s = s + "0";
				carry = 0;				
			}
			else if (tmp == 1)
			{
				s = s + "1";
				carry = 0;				
			}
			else if (tmp == 2)
			{
				s = s + "0";
				carry = 1;				
			}
			else //if (tmp == 3)
			{
				s = s + "1";
				carry = 1;				
			}
        }
        if (carry)
            s += '1';

        reverse(s.begin(), s.end());
        return s;
    }
};

#if 0 //MYDEBUG
int main()
{
    Solution* sol = new Solution();

    assert("100" == sol->addBinary("11", "1"));
    assert("10" == sol->addBinary("1", "1"));
    assert("10101" == sol->addBinary("1010", "1011"));
    assert("1011" == sol->addBinary("1010", "1"));
    assert("11100" == sol->addBinary("10101", "111"));
    cout<<"PASS!!"<<endl;;
    
    free(sol);
}
#endif

File: 67.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        '''
        a: str
        b: str
        return: str
        '''
        lena = len(a)
        lenb = len(b)
        s = ''
        carry = 0
        vala = valb = 0
        tmp = 0
        while lena > 0 or lenb > 0:
            lena = lena - 1
            lenb = lenb - 1
            if lena >= 0:
                vala = int(a[lena])
            else:
                vala = 0
            if lenb >= 0:
                valb = int(b[lenb])
            else:
                valb = 0
            tmp = vala + valb + carry
            s = s + str(tmp % 2)
            carry = tmp >> 1
        
        if carry > 0:
            s = s + '1'
        
        s = s[::-1] #reverse string
        #print(s)
        return s

'''
if __name__ == "__main__":
    sol = Solution()
    assert "100" == sol.addBinary("11","1"),"TEST FAIL"
    assert "10" == sol.addBinary("1","1"),"TEST FAIL"
    assert "10101" == sol.addBinary("1010","1011"),"TEST FAIL"
    assert "1011" == sol.addBinary("1010","1"),"TEST FAIL"
    assert "11100" == sol.addBinary("10101","111"),"TEST FAIL"
    print("TEST PASS")
'''